association.go

```diff
@@ -1714,34 +1925,54 @@ func (a *Association) processSelectiveAck(selectiveAckChunk *chunkSelectiveAck)
 			//        packets that were retransmitted (and thus for which it is
 			//        ambiguous whether the reply was for the first instance of the
 			//        chunk or for a later instance)
-			if chunkPayload.nSent == 1 && sna32GTE(chunkPayload.tsn, a.minTSN2MeasureRTT) {
-				a.minTSN2MeasureRTT = a.myNextTSN
-				rtt := time.Since(chunkPayload.since).Seconds() * 1000.0
-				srtt := a.rtoMgr.setNewRTT(rtt)
-				a.srtt.Store(srtt)
-				a.log.Tracef("[%s] SACK: measured-rtt=%f srtt=%f new-rto=%f",
-					a.name, rtt, srtt, a.rtoMgr.getRTO())
+			if sna32GTE(chunkPayload.tsn, a.minTSN2MeasureRTT) {
+				// Only original transmissions for classic RTT measurement (Karn's rule)
+				if chunkPayload.nSent == 1 {
+					a.minTSN2MeasureRTT = a.myNextTSN
+					rtt := now.Sub(chunkPayload.since).Seconds() * 1000.0
+					srtt := a.rtoMgr.setNewRTT(rtt)
+					a.srtt.Store(srtt)
+
+					// use a window to determine minRtt instead of a global min
+					// as the RTT can fluctuate, which can cause problems if going from a
+					// high RTT to a low RTT.
+					a.rackMinRTTWnd.Push(now, now.Sub(chunkPayload.since))
+
+					a.log.Tracef("[%s] SACK: measured-rtt=%f srtt=%f new-rto=%f",
+						a.name, rtt, srtt, a.rtoMgr.getRTO())
+				}
+
+				// RFC 8985 (RACK) sec 5.2: RACK.segment is the most recently sent
+				// segment that has been delivered, including retransmissions.
+				if chunkPayload.since.After(newestDeliveredSendTime) {
+					newestDeliveredSendTime = chunkPayload.since
+					newestDeliveredOrigTSN = chunkPayload.tsn
+					deliveredFound = true
+				}
 			}
```

Comment:
I found through SCP that this cases regression in RTT tests with dynamic re-order and 3% dropped packets, after looking in pprof and some debugging, I think this was a typo.
> In reality, as an algorithm, RACK does not arm a timer for every segment sent because it's not necessary. Instead, the sender records the most recent transmission time of every data segment sent, including retransmissions. For each ACK received, the sender calculates the latest RTT measurement (if eligible) and adjusts the expiration time of every segment sent but not yet delivered. If a segment has expired, RACK marks it as lost.
> https://datatracker.ietf.org/doc/html/rfc8985

```suggestion
			}

			// RFC 8985 (RACK) sec 5.2: RACK.segment is the most recently sent
			// segment that has been delivered, including retransmissions.
			if chunkPayload.since.After(newestDeliveredSendTime) {
				newestDeliveredSendTime = chunkPayload.since
				newestDeliveredOrigTSN = chunkPayload.tsn
				deliveredFound = true
			}
```

Reply:
Thanks for catching this, I fixed it in both places 😅